sql/api/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -784,35 +787,147 @@ abstract class SparkSession extends Serializable with Closeable {
 object SparkSession extends SparkSessionCompanion {
   type Session = SparkSession
 
-  private[this] val companion: SparkSessionCompanion = {
-    val cls = SparkClassUtils.classForName("org.apache.spark.sql.SparkSession")
+  // Config key/values used to set the SparkSession API mode.
+  val API_MODE_KEY: String = "spark.api.mode"
+  val API_MODE_CLASSIC: String = "classic"
+  val API_MODE_CONNECT: String = "connect"
+
+  // Implementation specific companions
+  private lazy val CLASSIC_COMPANION = lookupCompanion(
+    "org.apache.spark.sql.classic.SparkSession")
+  private lazy val CONNECT_COMPANION = lookupCompanion(
+    "org.apache.spark.sql.connect.SparkSession")
+  private def DEFAULT_COMPANION =
+    Try(CLASSIC_COMPANION).orElse(Try(CONNECT_COMPANION)).getOrElse {
+      throw new IllegalStateException(
+        "Cannot find an SparkSession implementation on the Classpath.")
+    }
+
+  private[this] def lookupCompanion(name: String): SparkSessionCompanion = {
+    val cls = SparkClassUtils.classForName(name)
     val mirror = scala.reflect.runtime.currentMirror
     val module = mirror.classSymbol(cls).companion.asModule
     mirror.reflectModule(module).instance.asInstanceOf[SparkSessionCompanion]
   }
 
   /** @inheritdoc */
-  override def builder(): SparkSessionBuilder = companion.builder()
+  override def builder(): Builder = new Builder
 
   /** @inheritdoc */
-  override def setActiveSession(session: SparkSession): Unit =
-    companion.setActiveSession(session.asInstanceOf[companion.Session])
+  override def setActiveSession(session: SparkSession): Unit = super.setActiveSession(session)
 
   /** @inheritdoc */
-  override def clearActiveSession(): Unit = companion.clearActiveSession()
+  override def setDefaultSession(session: SparkSession): Unit = super.setDefaultSession(session)
 
   /** @inheritdoc */
-  override def setDefaultSession(session: SparkSession): Unit =
-    companion.setDefaultSession(session.asInstanceOf[companion.Session])
+  override def getActiveSession: Option[SparkSession] = super.getActiveSession
 
   /** @inheritdoc */
-  override def clearDefaultSession(): Unit = companion.clearDefaultSession()
+  override def getDefaultSession: Option[SparkSession] = super.getDefaultSession
 
-  /** @inheritdoc */
-  override def getActiveSession: Option[SparkSession] = companion.getActiveSession
+  class Builder extends SparkSessionBuilder {
```

Comment:
This is by far and large the most important change. This builder supports both implementations.